src/target/riscv/riscv-013_reg.c

```diff
@@ -85,33 +87,247 @@ static int riscv013_reg_set(struct reg *reg, uint8_t *buf)
+	int res = riscv013_execute_abstract_command(target, command, &cmderr);
+	if (res == ERROR_OK) {
+		r->xlen = 64;
+		return ERROR_OK;
+	}
+	if (res == ERROR_TIMEOUT_REACHED)
 		return ERROR_FAIL;
+	r->xlen = 32;
+
+	return ERROR_OK;
+}
+
+static int examine_vlenb(struct target *target)
+{
+	RISCV_INFO(r);
+
+	/* Reading "vlenb" requires "mstatus.vs" to be set, so "mstatus" should
+	 * be accessible.*/
+	int res = init_cache_entry(target, GDB_REGNO_MSTATUS);
+	if (res != ERROR_OK)
+		return res;
+
+	res = assume_reg_exist(target, GDB_REGNO_VLENB);
+	if (res != ERROR_OK)
+		return res;
+
+	riscv_reg_t vlenb_val;
+	if (riscv_reg_get(target, &vlenb_val, GDB_REGNO_VLENB) != ERROR_OK) {
+		if (riscv_supports_extension(target, 'V'))
+			LOG_TARGET_WARNING(target, "Couldn't read vlenb; vector register access won't work.");
+		r->vlenb = 0;
+		return riscv_reg_impl_set_exist(target, GDB_REGNO_VLENB, false);
+	}
+	/* As defined by RISC-V V extension specification:
+	 * https://github.com/riscv/riscv-v-spec/blob/2f68ef7256d6ec53e4d2bd7cb12862f406d64e34/v-spec.adoc?plain=1#L67-L72 */
+	const unsigned int vlen_max = 65536;
+	const unsigned int vlenb_max = vlen_max / 8;
+	if (vlenb_val > vlenb_max) {
+		LOG_TARGET_WARNING(target, "'vlenb == %" PRIu64
+				"' is greater than maximum allowed by specification (%u); vector register access won't work.",
+				vlenb_val, vlenb_max);
+		r->vlenb = 0;
+		return ERROR_OK;
+	}
+	assert(vlenb_max <= UINT_MAX);
+	r->vlenb = (unsigned int)vlenb_val;
+
+	LOG_TARGET_INFO(target, "Vector support with vlenb=%u", r->vlenb);
+	return ERROR_OK;
+}
+
+enum misa_mxl {
+	MISA_MXL_INVALID = 0,
+	MISA_MXL_32 = 1,
+	MISA_MXL_64 = 2,
+	MISA_MXL_128 = 3
+};
+
+enum misa_mxl xlen_to_mxl(unsigned int xlen)
+{
+	switch (xlen) {
+	case 32:
+		return MISA_MXL_32;
+	case 64:
+		return MISA_MXL_64;
+	case 128:
+		return MISA_MXL_128;
+	}
+	return MISA_MXL_INVALID;
+}
+
+static int check_misa_mxl(const struct target *target)
+{
+	RISCV_INFO(r);
+
+	if (r->misa == 0) {
+		LOG_TARGET_WARNING(target, "'misa' register is read as zero."
+				"OpenOCD will not be able to determine some hart's capabilities.");
+		return ERROR_OK;
+	}
+	const unsigned int xlen = riscv_xlen(target);
+	assert(xlen <= sizeof(riscv_reg_t) * CHAR_BIT);
+	assert(xlen >= 2);
+	riscv_reg_t misa_mxl_field = (riscv_reg_t)0x3 << (xlen - 2);
+	const unsigned int mxl = get_field(r->misa, misa_mxl_field);
+	/* RISC-V Debug Spec. requires "DXLEN" to be the widest. */
+	if (mxl == MISA_MXL_INVALID || xlen_to_mxl(xlen) < mxl) {
+		LOG_TARGET_ERROR(target,
+				"'misa.mxl' field's value of %u does not "
+				" coincide with 'xlen' of %u.",
+				mxl, xlen);
+		return ERROR_FAIL;
+	}
+	/** NOTE:
+	 * The value of "misa.mxl" may stil not coincide with "xlen".
+	 * "misa[26:XLEN-3]" bits are marked as WIRI in at least version 1.10
+	 * of RISC-V Priveleged Spec. Therefore, if "xlen" is erroneously
+	 * assumed to be 32 when it actually is 64, "actual_mxl" will be read
```

Comment:
> if "xlen" is erroneously assumed to be 32 when it actually is 64 (...)

How could that happen? Do you mean some sort of a sporadic issue that could interfere with the 64-bit read of s0 (in examine()), making OpenOCD believe that the target is a 32-bit one?

Reply:
@JanMatCodasip yes. We sometimes observe such situations. For example if the system was under reset on some designs really weird things may happen.

Reply:
Thanks for the explanation.